packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.tsx

```diff
@@ -479,33 +476,48 @@ export class ContextualMenu extends BaseComponent<IContextualMenuProps, IContext
   private _onItemSubMenuExpand(item: IContextualMenuItem, target: HTMLElement) {
     if (this.state.expandedMenuItemKey !== item.key) {
 
-      if (this.state.submenuProps) {
+      if (this.state.expandedMenuItemKey) {
         this._onSubMenuDismiss();
       }
 
-      let submenuProps = {
-        items: getSubmenuItems(item),
-        target: target,
-        onDismiss: this._onSubMenuDismiss,
-        isSubMenu: true,
-        id: this.state.subMenuId,
-        shouldFocusOnMount: true,
-        directionalHint: getRTL() ? DirectionalHint.leftTopEdge : DirectionalHint.rightTopEdge,
-        className: this.props.className,
-        gapSpace: 0
-      };
-
-      if (item.subMenuProps) {
-        assign(submenuProps, item.subMenuProps);
-      }
-
       this.setState({
         expandedMenuItemKey: item.key,
-        submenuProps: submenuProps,
+        submenuTarget: target
       });
     }
   }
 
+  private _getSubmenuProps() {
+    let { submenuTarget, expandedMenuItemKey } = this.state;
+    let item = this._findItemByKey(expandedMenuItemKey);
+    let submenuProps = {
+      items: getSubmenuItems(item),
+      target: submenuTarget,
+      onDismiss: this._onSubMenuDismiss,
+      isSubMenu: true,
+      id: this.state.subMenuId,
+      shouldFocusOnMount: true,
+      directionalHint: getRTL() ? DirectionalHint.leftTopEdge : DirectionalHint.rightTopEdge,
+      className: this.props.className,
+      gapSpace: 0
+    };
+
+    if (item.subMenuProps) {
+      assign(submenuProps, item.subMenuProps);
+    }
+    return submenuProps;
+  }
+
+  private _findItemByKey(key: string): IContextualMenuItem {
+    let { items } = this.props;
+    for (let i = 0; i < this.props.items.length; i++) {
```

Comment:
It looks like `item` in this case equals the index of item. So I could rewrite it as 
``` 
for(const item in items) {
  let it=items[item] ... ```